### PR TITLE
 File Deletion using nonce, file renaming,  Video File Handling

### PR DESCRIPTION
**Pull request by Wojbie**
- File Deletion using nonce, file renaming
If ``uploadDeleteLink`` setting is false then file deletion is only possible for logged in user. If it's true then any user is allowed to remove if they provide correct nonce.
Also only if that setting is true the ``deleteUrl`` json key is added to returned url on upload that informs unloader what deletionUrl is.
Also added success page to offset error page. Provided some basic custom name functions for end user to use.
Did some basic GUI work so there are buttons for both rename and delete everywhere. Can't get it to look correctly in list/links pages - little offset.
For security reasons no path input is allowed from user - all files that are to be changed are selected using their nonce.

 - Video File Handling to Gallery
Cause why not have it load fist frame from video files? It makes sense in current world where people start sharing mp4 and webm. Preloading is set to metadata only so it should not waste anyone data plan while still getting first frame of video as thumbnail.

 - Examples for custom-name functions
Dunno - was bored - seemed like good idea at time.

 - Saner statCache invalidation.
So that every time something new is added that need's to cache the data it don't need it's own handling code. 
Warning - This change will cause old statCache files to be re-created. Every time statCache format is changed version number should be changed too to ensure old formatted statCache is not used by new code.

 - Additional Uploading handling
Fixed small bug with Upload that allowed uploading empty files. Also made it so that if file is dragged over `Upload` link on any of the pages then page will switch to uploading one.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -221,29 +271,42 @@ router.post("/upload", (req, res) => {
...
+			if (err) error(req, res, "Upload failed."); return console.log(JSON.stringify(err));
+			let nonce = generateNonce(`${config.imagePath}/${name}`)
 			name = "l/" + name;
-			res.json({
-				ok: true,
-				url: `${config.url.replace(/\/?$/, "/")}${name}`
-			});
+			
+			if (req.body.online === "yes") {
+				success(req, res, `URL shortened to <a href="${config.url}${name}">"${config.url}${name}"</a>` );
```
Comment (Lemmmy) [inline index.js]: Potential XSS

File: views/layouts/main.hbs [hunk 2]
```diff
@@ -159,7 +176,7 @@
 							window.location.href = result.url ;
 						}
 						else {
-							$("body").html('<div class="ui middle aligned center aligned grid"><div class="column"><div class="ui centered green card"><div class="content"><h1 class="header" style="color: #229A33 !important; font-size: 175%">Success</h1></div><p class="ui content text" style="font-size: 120%">'+result.url+'</p></div></div></div>')
+							$("body").html('<div class="ui middle aligned center aligned grid"><div class="column"><div class="ui centered green card"><div class="content"><h1 class="header" style="color: #229A33 !important; font-size: 175%">Success</h1></div><p class="ui content text" style="font-size: 120%">'+result.url+'</p><p class="ui content text" style="font-size: 120%"><button onClick="window.location.reload(true)" type="button">Go Back</button></p></div></div></div>')
```
Comment (Lemmmy) [inline views/layouts/main.hbs]: Potential XSS - use jquery templates or dynamically create elements and use $.text(result.url)

File: views/error.hbs [hunk 3]
```diff
@@ -7,6 +7,9 @@
 			<p class="ui content text" style="font-size: 120%">
 				{{ errorText }}
 			</p>
+			<p class="ui content text" style="font-size: 120%">
+				<button onClick="window.location.href = document.referrer" type="button">Go Back</button>
```
Comment (Lemmmy) [inline views/error.hbs]: could this be made into a semantic ui button?

File: views/layouts/main.hbs [hunk 4]
```diff
@@ -121,7 +138,7 @@
 			});
 			
 			function errorMessage(err) {
-				$("body").html('<div class="ui middle aligned center aligned grid"><div class="column"><div class="ui centered green card"><div class="content"><h1 class="header" style="color: #DB2828 !important; font-size: 175%">Error</h1></div><p class="ui content text" style="font-size: 120%">'+err+'</p></div></div></div>')
+				$("body").html('<div class="ui middle aligned center aligned grid"><div class="column"><div class="ui centered green card"><div class="content"><h1 class="header" style="color: #DB2828 !important; font-size: 175%">Error</h1></div><p class="ui content text" style="font-size: 120%">'+err+'</p><p class="ui content text" style="font-size: 120%"><button onClick="window.location.reload(true)" type="button">Go Back</button></p></div></div></div>')
```
Comment (Wojbie) [inline views/layouts/main.hbs]: Now that I am a bit saner this looks like xss too.

Comment (Lemmmy) [inline views/layouts/main.hbs]: semantic ui this button too?

File: views/success.hbs [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+	<div class="column">
+		<div class="ui centered green card">
+			<div class="content">
+				<h1 class="header" style="color: #229A33 !important; font-size: 175%">Success</h1>
+			</div>
+			<p class="ui content text" style="font-size: 120%">
+				{{ successText }}
+			</p>
+			<p class="ui content text" style="font-size: 120%">
+				<button onClick="window.location.href = document.referrer" type="button">Go Back</button>
```
Comment (Lemmmy) [inline views/success.hbs]: semantic ui this button too?

File: index.js [hunk 6]
```diff
@@ -41,11 +48,21 @@ if (fs.existsSync("custom-name.js")) {
...
+	else {
+		for (let key in statCache) {
+			if (!statCache.hasOwnProperty(key) || key == "version") continue;
+			statCache[key].mtime = new Date(statCache[key].mtimeSave);
+		}
+	}
+}
+
+if (fs.existsSync("nonces.json")) {
+	nonces = JSON.parse(fs.readFileSync("nonces.json"));
```
Comment (Lignum) [inline index.js]: Wouldn't it be better if this was in a database or something? This sounds like a job for Redis.

Comment (Lemmmy) [reply index.js]: redis doesn't have persistence by default - the whole point of this is persistence, on the real shitty.download instance it takes about 3 minutes to fs.stat all the files

**Comment by Lemmmy** [comment]
also my nonce JSON somehow became corrupted during testing?

**Comment by Wojbie** [comment]
Opse. Sorry for accidental fake merge in there.

**Comment by Lemmmy** [comment]
you ready to merge thsi?

**Comment by Wojbie** [comment]
Looks fine from my side. Are you satisfied with it?

**Comment by Lemmmy** [comment]
yep!

**Pull request merged**